// harbors2/module/cluster.js
/**
 * The process of cluster module
 */
var cluster = require('cluster');
var http = require('http');
var fs = require('fs');
var _m_path = require('../module/path');
var _m_request = require('../module/request');
var _m_server = require('../module/server');

/**
 * master process
 */
exports.master = function(){

    var message = function(msg){

    };

    var createWorker = function(){
        var worker = cluster.fork();
        worker.on('exit', function(code, signal) {
            //重启worker（此时进程id会增加1，以示和原错误关闭进程区别）
            !worker.suicide && (console.log('Separation of new sub process process...'),setTimeout(createWorker,2000));
        });
    };

    for(var i=0;i<harbors._process;i++){
        createWorker();
    }

};

/**
 * worker process
 * @param {Function} callback
 */
exports.worker = function(callback){
    http.createServer(function(req,res){
        callback(req,res);
    }).listen(harbors._listen);
};

exports.server = function(req,res){
    var p = _m_path.analysis(req);
    if(p.File=='favicon.ico') return false;
//    console.log(p);
    /*{
     vhostName: 'server',
     Dir: '/',
     File: 'index.html',
     Extname: '.html',
     Module: 'index_.js'
     }*/
//    console.log(p.Dir + p.Module)
    var stM = !!(p.Extname!='' && fs.existsSync(p.Dir + p.File));
    var dyM = fs.existsSync(p.Dir + p.Module);
    var request = _m_request(req,res,p,harbors.config[p.vhostName]);
//    console.log(stM,dyM);
    try{
        stM ?
            request.display(request.path.File) :
            dyM ?
                _m_server.dyM(request) :
                request.notFound();
    }catch(err){
        console.log(err);
        request.error();
    }
};
